Make new 7 interactive

BillWagner opened the pull request:
Fixes #5432 
Fixes #7506 

Contributes to #8206

Internal review links:
[What's new in C# 7](https://review.docs.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-7?branch=pr-en-us-11442)
[Explore C# 7](https://review.docs.microsoft.com/en-us/dotnet/csharp/tutorials/exploration/csharp-7?branch=pr-en-us-11442)

BillWagner committed:
finish trimming "what's new"

BillWagner committed:
add interactive tutorial

BillWagner committed:
fix the build errors

rpetrusha commented on docs/csharp/tutorials/exploration/csharp-7.yml:
```diff
@@ -0,0 +1,536 @@
+    > on platforms that do not include the types.
+    >
+    > This is similar to other language features that rely on types
+    > delivered in the framework. Example include `async` and `await`
```
```suggestion
    > delivered in the framework. Examples include `async` and `await`
```

rpetrusha commented on docs/csharp/tutorials/exploration/csharp-7.yml:
```diff
@@ -0,0 +1,536 @@
+    > as .NET is becoming more platform independent. The .NET Framework
+    > may not always ship on the same cadence as the language compiler. When new language
+    > features rely on new types, those types will be available as NuGet packages when
+    > the language features ship. As these new types get added to the .NET Standard
```
```suggestion
    > the language features ship. As these new types are added to the .NET Standard
```

rpetrusha commented on docs/csharp/tutorials/exploration/csharp-7.yml:
```diff
@@ -0,0 +1,536 @@
+    > may not always ship on the same cadence as the language compiler. When new language
+    > features rely on new types, those types will be available as NuGet packages when
+    > the language features ship. As these new types get added to the .NET Standard
+    > API and delivered as part of the framework, the NuGet package requirement will
```
```suggestion
    > Library and delivered as part of the framework, the NuGet package requirement will
```
framework, or .NET Framework?

rpetrusha commented on docs/csharp/tutorials/exploration/csharp-7.yml:
```diff
@@ -0,0 +1,536 @@
+    > API and delivered as part of the framework, the NuGet package requirement will
+    > be removed.
+
+    C# provides a rich syntax for classes and structs that is used to explain
```
```suggestion
    C# provides a rich syntax for classes and structs that explains
```

rpetrusha commented on docs/csharp/tutorials/exploration/csharp-7.yml:
```diff
@@ -0,0 +1,536 @@
+    C# provides a rich syntax for classes and structs that is used to explain
+    your design intent. But sometimes that rich syntax requires extra
+    work with minimal benefit. You may often write methods that need a simple
+    structure containing more than one data element. To support these scenarios
```
```suggestion
    structure containing more than one data element. To support these scenarios,
```

rpetrusha commented on docs/csharp/tutorials/exploration/csharp-7.yml:
```diff
@@ -0,0 +1,536 @@
+
+    The `namedLetters` tuple contains fields referred to as `Alpha` and
+    `Beta`. Those names exist only at compile time and are not preserved
+    for example when inspecting the tuple using reflection at runtime.
```
```suggestion
when inspecting the tuple using reflection at runtime, for example.
```

BillWagner committed:
build errors, round 2

rpetrusha commented on docs/csharp/tutorials/exploration/csharp-7.yml:
```diff
@@ -0,0 +1,536 @@
+    [!code-csharp[NamedTupleConflict](~/samples/snippets/csharp/new-in-7/program.cs#07_NamedTupleConflict "Named tuple conflict")]
+
+    The line above generates a warning, `CS8123`, telling you that the names on the right
+    side of the assignment, `Alpha` and `Beta` are ignored because they conflict
```
```suggestion
    side of the assignment, `Alpha` and `Beta`, are ignored because they conflict
```
Incidentally, running this interactively just generates a "No output" message, not a compiler warning.

rpetrusha commented on docs/csharp/tutorials/exploration/csharp-7.yml:
```diff
@@ -0,0 +1,536 @@
+    most useful as return types for `private` and `internal` methods. Tuples
+    provide a simple syntax for those methods to return multiple discrete values:
+    You save the work of authoring a `class` or a `struct` that
+    defines the type returned. There is no need for creating a new type.
```
```suggestion
    defines the type returned. There is no need to create a new type.
```

rpetrusha commented on docs/csharp/tutorials/exploration/csharp-7.yml:
```diff
@@ -0,0 +1,536 @@
+    defines the type returned. There is no need for creating a new type.
+
+    Creating a tuple is more efficient and more productive.
+    It is a simpler, lightweight syntax to define a data structure that carries
```
```suggestion
    It has a simpler, lightweight syntax to define a data structure that carries
```

rpetrusha commented on docs/csharp/tutorials/exploration/csharp-7.yml:
```diff
@@ -0,0 +1,536 @@
+
+    * You save the work of authoring a `class` or a `struct` that defines the type returned. 
+    * You do not need to create new type.
+    * The language enhancements removes the need to call the <xref:System.Tuple.Create``1(``0)> methods.
```
```suggestion
    * The language enhancements removes the need to call the <xref:System.Tuple.Create``1(``0)?displayProperty=nameWithType> methods.
```

rpetrusha commented on docs/csharp/tutorials/exploration/csharp-7.yml:
````diff
@@ -0,0 +1,536 @@
+    return result;
+    ```
+
+- title: Tuples create light-weight data structures
````
You might consider adding output to some of the tuples examples so that they can run interactively.

rpetrusha commented on docs/csharp/tutorials/exploration/csharp-7.yml:
```diff
@@ -0,0 +1,536 @@
+- title: Use discards to select needed members
+  durationInMinutes: 2
+  content: |
+    Often when deconstructing a tuple or calling a method with `out` parameters, you're forced to define a variable whose value you don't care about and don't intend to use. C# adds support for *discards* to handle this scenario. A discard is a write-only variable whose name is `_` (the underscore character); you can assign all of the values that you intend to discard to the single variable. A di ...
```
```suggestion
    Often when deconstructing a tuple or calling a method with `out` parameters, you're forced to define a variable whose value you don't care about and don't intend to use. C# adds support for *discards* to handle this scenario. A discard is a write-only variable whose name is `_` (the underscore character); you can assign all of the values that you intend to discard to the single variable. A discard is like an unassigned variable; apart from the assignment, the discard can't be used in code.
```
There is no statement in a formal sense; it's a predefined variable.

rpetrusha commented on docs/csharp/tutorials/exploration/csharp-7.yml:
```diff
@@ -0,0 +1,536 @@
+
+    * As a standalone identifier when you want to explicitly identify the value of an assignment as a discard.
+
+    The following example defines a `QueryCityDataForYears` method that returns a 6-tuple that contains a data for a city for two different years. The method call in the example is concerned only with the two population values returned by the method and so treats the remaining values in the tuple as discards when it deconstructs the tuple.
```
```suggestion
    The following example defines a `QueryCityDataForYears` method that returns a 6-tuple that contains data for a city for two different years. The method call in the example is concerned only with the two population values returned by the method and so treats the remaining values in the tuple as discards when it deconstructs the tuple.
```

rpetrusha commented on docs/csharp/tutorials/exploration/csharp-7.yml:
```diff
@@ -0,0 +1,536 @@
+  content: |
+    *Pattern matching* is a feature that allows you to implement method dispatch on
+    properties other than the type of an object. You're probably already familiar
+    with method dispatch based on the type of an object. In Object Oriented programming,
```
```suggestion
    with method dispatch based on the type of an object. In object oriented programming,
```

rpetrusha commented on docs/csharp/tutorials/exploration/csharp-7.yml:
```diff
@@ -0,0 +1,536 @@
+    properties other than the type of an object. You're probably already familiar
+    with method dispatch based on the type of an object. In Object Oriented programming,
+    virtual and override methods provide language syntax to implement method dispatching
+    based on an object's type. Base and Derived classes provide different implementations. 
```
```suggestion
    based on an object's type. Base and derived classes provide different implementations. 
```

rpetrusha commented on docs/csharp/tutorials/exploration/csharp-7.yml:
```diff
@@ -0,0 +1,536 @@
+
+    Pattern matching supports `is` expressions and `switch` expressions. Each
+    enables inspecting an object and its properties to determine if that object
+    satisfies the sought pattern. You use the `when` keyword to specify additional
```
```suggestion
    satisfies the desired pattern. You use the `when` keyword to specify additional
```

BillWagner committed:
one more build warning

rpetrusha commented on docs/csharp/tutorials/exploration/csharp-7.yml:
```diff
@@ -0,0 +1,536 @@
+  content: |
+    This feature enables algorithms that use and return references
+    to variables defined elsewhere. One example is working with
+    large matrices, and finding a single location with certain
```
```suggestion
    large matrices and finding a single location with certain
```

rpetrusha commented on docs/csharp/tutorials/exploration/csharp-7.yml:
```diff
@@ -0,0 +1,536 @@
+    [!code-csharp[FindReturningIndices](~/samples/snippets/csharp/new-in-7/MatrixSearch.cs#20_FindReturningIndices "Find returning indices")]
+
+    There are many issues with this code. First of all, it's a public
+    method that's returning a tuple. The language supports this, but
```
```suggestion
    method that returns a tuple. The language supports this, but
```

rpetrusha commented on docs/csharp/tutorials/exploration/csharp-7.yml:
```diff
@@ -0,0 +1,536 @@
+
+    There are many issues with this code. First of all, it's a public
+    method that's returning a tuple. The language supports this, but
+    user defined types (either classes or structs) are preferred
```
```suggestion
    user-defined types (either classes or structs) are preferred
```

rpetrusha commented on docs/csharp/tutorials/exploration/csharp-7.yml:
```diff
@@ -0,0 +1,536 @@
+    user defined types (either classes or structs) are preferred
+    for public APIs.
+
+    Second, this method is returning the indices to the item in the matrix.
```
```suggestion
    Second, this method returns the indices to the item in the matrix.
```

rpetrusha commented on docs/csharp/tutorials/exploration/csharp-7.yml:
```diff
@@ -0,0 +1,536 @@
+    Let's walk through a series of changes to demonstrate the ref local feature
+    and show how to create a method that returns a reference to internal storage.
+    Along the way, you'll learn the rules of the ref return and ref local feature that
+    protects you from accidentally misusing it.
```
```suggestion
    protect you from accidentally misusing it.
```

rpetrusha commented on docs/csharp/tutorials/exploration/csharp-7.yml:
```diff
@@ -0,0 +1,536 @@
+    [!code-csharp[FindReturningRef](~/samples/snippets/csharp/new-in-7/MatrixSearch.cs#22_FindReturningRef "Find returning by reference")]
+
+    Now that the method returns a reference to the integer value in the
+    matrix, you need to modify where it's called.  The `var` declaration
```
where it's called --> the call site?

rpetrusha commented on docs/csharp/tutorials/exploration/csharp-7.yml:
```diff
@@ -0,0 +1,536 @@
+
+    [!code-csharp[AssignRefReturn](~/samples/snippets/csharp/new-in-7/program.cs#24_AssignRefReturn "Assign ref return")]
+
+    Now, the second `WriteLine` statement in the example above will print 
```
```suggestion
    Now, the second `WriteLine` statement in the example above prints 
```

rpetrusha commented on docs/csharp/tutorials/exploration/csharp-7.yml:
```diff
@@ -0,0 +1,536 @@
+    out the value `24`, indicating that the storage in the matrix has been
+    modified. The local variable has been declared with the `ref` modifier,
+    and it will take a `ref` return. You must initialize a `ref` variable when
+    it is declared, you cannot split the declaration and the initialization.
```
```suggestion
    it is declared; you cannot split the declaration and the initialization.
```

rpetrusha commented on docs/csharp/tutorials/exploration/csharp-7.yml:
```diff
@@ -0,0 +1,536 @@
+
+    For more information, see the [ref keyword](../../language-reference/keywords/ref.md) article.
+
+- title: Minimize access to code with local functions
```
You might consider moving local functions to earlier. You can then use local functions in some of your ref returns and tuple examples in making them interactive.

BillWagner replied:
I kept the same order, but hid the local functions behind the try.net scaffolding.

rpetrusha commented on docs/csharp/tutorials/exploration/csharp-7.yml:
```diff
@@ -0,0 +1,536 @@
+    Many designs for classes include methods that are called from only
+    one location. These additional private methods keep each method small
+    and focused. However, they can make it harder to understand a class
+    when reading it the first time. These methods must be understood
```
```suggestion
    when reading it the first time, since these methods are understood
```

rpetrusha commented on docs/csharp/tutorials/exploration/csharp-7.yml:
```diff
@@ -0,0 +1,536 @@
+    > [!NOTE]
+    > Some of the designs that are supported by local functions
+    > could also be accomplished using *lambda expressions*. Those
+    > interested can [read more about the differences](../../local-functions-vs-lambdas.md)
```
```suggestion
    > interested can [read more about the differences](../../local-functions-vs-lambdas.md).
```

rpetrusha commented on docs/csharp/tutorials/exploration/csharp-7.yml:
```diff
@@ -0,0 +1,536 @@
+    > could also be accomplished using *lambda expressions*. Those
+    > interested can [read more about the differences](../../local-functions-vs-lambdas.md)
+
+- title: Throw expressions more concise error reporting
```
```suggestion
- title: Throw expressions allow more concise error reporting
```

rpetrusha reviewed (changes requested):
I've left a number of suggestions and comments, @BillWagner. I've requested changes rather than approved, since I'd like to review this again once it becomes more interactive.

Commit:
Apply suggestions from code review

A very thorough early review from Ron

Co-Authored-By: BillWagner <wiwagn@microsoft.com>

BillWagner committed:
can't use a : in the content.

BillWagner committed:
still experimenting with YML format issues.

BillWagner committed:
experimenting again

BillWagner committed:
Notes and a test build.

BillWagner committed:
rework the interactive tutorial

BillWagner committed:
typo / fix broken link

pkulikov commented on docs/csharp/tutorials/exploration/csharp-7.yml:
```diff
@@ -0,0 +1,263 @@
+### YamlMime:Tutorial
+title: Explore C# 7 - C# interactive C# tutorial
```
```suggestion
title: Explore C# 7.0 - C# interactive tutorial
```

pkulikov commented on docs/csharp/tutorials/exploration/csharp-7.yml:
```diff
@@ -0,0 +1,263 @@
+### YamlMime:Tutorial
+title: Explore C# 7 - C# interactive C# tutorial
+metadata:
+  title: Explore C# 7 - Try the new features in C# 6 interactively, using your browser
```
```suggestion
  title: Explore C# 7.0 - Try the new features in C# 7.0 interactively, using your browser
```

pkulikov commented on docs/csharp/tutorials/exploration/csharp-7.yml:
```diff
@@ -0,0 +1,263 @@
+title: Explore C# 7 - C# interactive C# tutorial
+metadata:
+  title: Explore C# 7 - Try the new features in C# 6 interactively, using your browser
+  description: In this tutorial, you'll use your browser to explore C# 6 interactively. You'll explore the new idioms you can use with C# 6 that enable more concise and readable code.
```
```suggestion
  description: In this tutorial, you'll use your browser to explore C# 7.0 interactively. You'll explore the new idioms you can use with C# 7.0 that enable more concise and readable code.
```

pkulikov commented on docs/csharp/tutorials/exploration/csharp-7.yml:
```diff
@@ -0,0 +1,263 @@
+items:
+- durationInMinutes: 1
+  content: |
+    This tutorial lets you explore C# 7 features interactively, using your browser to write C# and see the results of compiling and running your code. It contains a series of lessons that modify earlier C# practices to use newer, more concise C# 7 features. The rest of this article provides an overview of each of these features, with a link to explore each feature.
```
```suggestion
    This tutorial lets you explore C# 7.0 features interactively, using your browser to write C# and see the results of compiling and running your code. It contains a series of lessons that modify earlier C# practices to use newer, more concise C# 7.0 features. The rest of this article provides an overview of each of these features, with a link to explore each feature.
```

pkulikov commented on docs/csharp/tutorials/exploration/csharp-7.yml:
````diff
@@ -0,0 +1,263 @@
+    return ref matrix[i,j];
+    ```
+
+    Now that the method returns a reference to the integer value in the matrix, you need to modify where it's called.  The `var` declaration means that `valItem` is now an `int` rather than a tuple. Change the calling code to following:
````
```suggestion
    Now that the method returns a reference to the integer value in the matrix, you need to modify where it's called. The `var` declaration means that `valItem` is now an `int` rather than a tuple. Change the calling code to following:
```

pkulikov commented on docs/toc.yml:
```diff
@@ -527,6 +527,8 @@
           href: csharp/tutorials/intro-to-csharp/introduction-to-classes.md
     - name: Explore C# 6
       href: csharp/tutorials/exploration/csharp-6.yml
+    - name: Explore C# 7
```
```suggestion
    - name: Explore C# 7.0
```

pkulikov reviewed (commented):
Nit-picking on C# versions. Please ignore most of suggestions if "C# 7" is by design: as an intention to cover C# 7.x features (to include those from C# 7.1...7.3 later)

pkulikov committed:
Apply suggestions from code review

Co-Authored-By: BillWagner <wiwagn@microsoft.com>

BillWagner commented:
closing and opening for a new build

BillWagner commented:
@rpetrusha I just updated the build, and gave it a quick look. It's ready for a fresh review.

BillWagner commented:
One of pattern matching samples behaves incorrectly in the try.net experience See dotnet/try#175

BillWagner committed:
Fix the samples for ref

These don't work unless the sample in the browser is a full program. Ref returns in snippets don't compile in the current scaffolding.

rpetrusha commented on docs/csharp/whats-new/csharp-7.md:
```diff
@@ -540,67 +303,46 @@ type, so using it means allocating an object. In cases where a
 write that number in binary:
 
-[!code-csharp[BinaryConstants](../../../samples/snippets/csharp/new-in-7/Program.cs#32_BinaryConstants "Binary constants")]
+[!code-csharp[ThousandSeparators](~/samples/snippets/csharp/new-in-7/Program.cs#33_ThousandSeparators "Thousands separators")]
```
```suggestion
[!code-csharp[ThousandSeparators](~/samples/snippets/csharp/new-in-7/Program.cs#ThousandSeparators "Thousands separators")]
```

rpetrusha commented on docs/csharp/whats-new/csharp-7.md:
```diff
@@ -540,67 +303,46 @@ type, so using it means allocating an object. In cases where a
-The digit separator can appear anywhere in the constant. For base 10
+number is written as a binary number. Binary numbers can get long, so it's often easier to see
+the bit patterns by introducing the `_` as a digit separator, as shown above in the binary constant. The digit separator can appear anywhere in the constant. For base 10
 numbers, it would be common to use it as a thousands separator:
```
```suggestion
numbers, it is common to use it as a thousands separator:
```

rpetrusha reviewed (approved):
This looks good, @BillWagner. Once the REPL is working, we should remember to revisit the interactive examples to make sure that they work. I've left a number of comments and suggestions, but feel free to merge when you want and, if necessary, address them later.

Commit:
Apply suggestions from code review

Co-Authored-By: BillWagner <wiwagn@microsoft.com>

Commit:
Apply suggestions from code review

Co-Authored-By: BillWagner <wiwagn@microsoft.com>

BillWagner committed:
respond to final feedback

BillWagner committed:
add note about try.net compiler issue
